pkg/sharding/apiserver/request.go

```diff
@@ -0,0 +1,74 @@
+package apiserver
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/client-go/rest"
+)
+
+// requestFor is a factory for *rest.Request from the source *http.Request, ensuring the source is not mutated
+func requestFor(req *http.Request) func(client *rest.RESTClient, mutators... func(runtime.Object) error) (*rest.Request, error) {
+	return func(client *rest.RESTClient, mutators... func(runtime.Object) error) (*rest.Request, error) {
+		body, err := ioutil.ReadAll(req.Body)
+		if err != nil {
+			return nil, err
+		}
+		if err := req.Body.Close(); err != nil {
+			return nil, err
+		}
+		if len(mutators) > 0 && len(body) > 0 {
+			// mutate the body as needed, using unstructured as our intermediate decoding
+			obj, _, err := unstructured.UnstructuredJSONScheme.Decode(body, nil, nil)
+			if err != nil {
+				return nil, fmt.Errorf("could not decode body as unstructured: %w", err)
+			}
+			for _, mutator := range mutators {
+				if err := mutator(obj); err != nil {
+					return nil, fmt.Errorf("could not mutate body: %w", err)
+				}
+			}
+			encoded := &bytes.Buffer{}
+			if err := unstructured.UnstructuredJSONScheme.Encode(obj, encoded); err != nil {
+				return nil, fmt.Errorf("could not encode mutated body: %w", err)
+			}
+			body = encoded.Bytes()
+		}
+		request := rest.NewRequest(client).
+			Verb(req.Method).
+			RequestURI(req.URL.RequestURI()).
+			SetHeader("X-Kubernetes-Sharded-Request", "true").
+			Body(body).Debug()
```

Comment:
Instead of a custom `Debug` method added to the forked repo, we could provide a custom RoundTripper defined in this repo.

Reply:
Pros/cons?